tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java

```diff
@@ -0,0 +1,216 @@
+package thredds.server.wfs;
+
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.List;
+import thredds.server.config.ThreddsConfig;
+
+/**
+ * A writer for a WFS compliant Geospatial XML given a print writer.
+ * Specifically writes the XML for WFS GetCapabilities requests.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class WFSGetCapabilitiesWriter {

```

Comment:
This class is highly stateful. Sometimes stateful designs are appropriate. They are usually harder to test than stateless designs. Given our goals and time constraints I do not want to mandate tests on this class, however I would like to check: have you considered a stateless or a less stateful design?

DZone has [a good article](https://dzone.com/articles/stateful-or-stateless-classes) on this topic if you are hungry for inspiration.

A few examples of state and possible alternatives:
 * `fileOutput`, a String that builds up the XML to be written to the `PrintWriter`. This could be removed in favor of writing results to the `PrintWriter` immediately, or by returning the results of each method as a String.
 * a list of WFSRequestTypes. This could be changed from an instance variable to a parameter on methods.
 * a list of WFSFeatures. This could be changed from an instance variable to a parameter on methods.

Just because these things could be done doesn't mean they should be done.  If you have weighed the pros and cons and feel a stateful design is appropriate, go for it. If you think a stateful design would be useful, but you would rather implement it in a subsequent PR, that is OK too.

Reply:
:) Very good points. Thanks for considering a stateless design. It's up to you!

Reply:
Thanks for looking through this pull request (and others as well) and providing suggestions on possible other methodologies that could be used!